OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,457 @@
+			public readonly uint	CompressedSize;
+			public readonly uint	DataOffset;
+
+			public readonly byte[]	MD5;
+			public readonly uint	NameOffset;
+			public readonly ushort	DirectoryIndex;
+			public readonly uint	LinkPrevious;
+
+			public readonly uint	LinkNext;
+			public readonly byte	LinkFlags;
+			public readonly ushort	Volume;
+			public readonly string	Filename;
+
+			public FileDescriptor(BinaryReader reader, long tableOffset)
+			{
+				Flags            = reader.ReadUInt16();
+				ExpandedSize     = reader.ReadUInt32();
+				/*    unknown   */ reader.ReadUInt32();
+				CompressedSize   = reader.ReadUInt32();
+
+				/*  unknown */ reader.ReadUInt32();
+				DataOffset   = reader.ReadUInt32();
+				/*  unknown */ reader.ReadUInt32();
+				MD5          = reader.ReadBytes(0x10);
+
+				/*   unknown  */ reader.ReadBytes(0x10);
+				NameOffset     = reader.ReadUInt32();
+				DirectoryIndex = reader.ReadUInt16();
+				/*   unknown  */ reader.ReadBytes(0xc);
+				LinkPrevious   = reader.ReadUInt32();
+				LinkNext       = reader.ReadUInt32();
+
+				LinkFlags   = reader.ReadByte();
+				Volume      = reader.ReadUInt16();
+				var posSave = reader.BaseStream.Position;
+
+				reader.BaseStream.Seek(tableOffset + NameOffset, SeekOrigin.Begin);
+				Filename = reader.BaseStream.ReadASCIIZ();
+				reader.BaseStream.Seek(posSave, SeekOrigin.Begin);
+			}
+		}
+#endregion
+		readonly Stream hdrFile;
+		CommonHeader commonHeader;
+		CabDescriptor cabDescriptor;
+		List<uint> directoryTable;
+		Dictionary<uint, string> directoryNames;
+		Dictionary<uint, FileDescriptor> fileDescriptors;
+		Dictionary<string, FileDescriptor> fileDict;
+		List<uint> fileGroupOffsets;
+		List<FileGroup> fileGroups;
+		int priority;
+		string commonName;
+		public int Priority { get { return priority; } }
+
+		public string Name { get { return commonName; } }
+
+		public InstallShieldCABExtractor(string hdrFilename, int priority = -1)
+		{
+			this.priority = priority;
+			hdrFile       = GlobalFileSystem.Open(hdrFilename);
+			var buff      = new List<char>(hdrFilename.Substring(0, hdrFilename.LastIndexOf('.')).ToCharArray());
+
+			for (int i = buff.Count - 1; char.IsNumber(buff[i]); --i)
+				buff.RemoveAt(i);
+
+			commonName    = new string(buff.ToArray());
+			var reader    = new BinaryReader(hdrFile);
+			var signature = reader.ReadUInt32();
+
+			if (signature != 0x28635349)
+				throw new InvalidDataException("Not an Installshield CAB package");
+
+			commonHeader     = new CommonHeader(reader);
+			cabDescriptor    = new CabDescriptor(reader, commonHeader);
+			/*    unknown   */ reader.ReadBytes(0xe);
+			fileGroupOffsets = new List<uint>();
+
+			for (uint i = MaxFileGroupCount; i > 0; --i)
+				fileGroupOffsets.Add(reader.ReadUInt32());
+
+			reader.BaseStream.Seek(commonHeader.CabDescriptorOffset + cabDescriptor.FileTableOffset, SeekOrigin.Begin);
+			directoryTable = new List<uint>();
+
+			for (uint i = cabDescriptor.DirectoryCount; i > 0; --i)
```

Comment:
`var` everywhere please, and I suggest incrementing up instead of down for consistency.